### DISCUSSION TITLE
Big-O complexity for complex datatypes

### DISCUSSION THREAD

**Discussion by NickCrews** [duckdb/duckdb#11077] [Q&A]
I am wondering what is the big-O complexity for various List and Map functions:
- `list_contains()`: O(n) using a loop, or O(1) using some precomputed hash method? Assuming we don't speculatively create a hashset, so I'm guessing this is O(n).
-  `list_extract()`: Assuming this is O(1)
- `list_intersect(m, n)`: [EDIT: looks like [O(m*n)](https://github.com/duckdb/duckdb/discussions/8178)] could be O(max(n, m)) (build a hashset of m then iterate through n), or could be O(m*n)
- `map_extract`: O(1) I would hope, but a naive implementation would just be a loop for O(n)
- etc, if you can speak in general what the underlying datastructures are

I ask because I am trying to [implement the vector dot product on sparse vectors of text using Map](https://github.com/ibis-project/ibisml/issues/26), and this only will work performantly if `list_intersect()` is O(max(n, m)) and `map_extract()` is O(1).

If I get an answer here, would you like me to open a PR adding this info to the [Map docs](https://duckdb.org/docs/sql/data_types/map) and [List docs](https://duckdb.org/docs/sql/data_types/list)?

**Comment by soerenwolfers** [comment]
Sadly, `list_intersect` seems to be `m*n` as you can verify with 
```sql
SELECT list_intersect(range(n), range(n / 2))
```

I think you should report that as an issue

**Reply by soerenwolfers** [reply]
Ah, this has already  been noticed in https://github.com/duckdb/duckdb/discussions/8178

**Reply by NickCrews** [reply]
Thanks! That issue looks useful, and that is a great microbenchmark for testing the complexity of things.

**Comment by NickCrews** [comment]
OK per the benchmarking clue above, I can benchmark using
```sql
-- create the data as a setup step so this is not included in the timing
CREATE OR REPLACE TABLE t as SELECT [RANDOM() for x in range(10e7::BIGINT)] as x;
```

and then timing `SELECT list_extract(x, 42542344) from t;` vs `SELECT list_extract(x, 34) from t;`, both of which appear to be instant. So list_extract seems like it is O(1).
